paperless-ngx: init at 1.6.0, update NixOS service

erikarvstedt opened the pull request:
#### Summary of this PR
- Add pkg `paperless-ngx`
- Remove pkg `paperless-ng` and add alias to `paperless-ngx`.
  `paperless-ngx` is backwards-compatible with `paperless-ng` and, so far, only contains minor improvements and bugfixes.
  There's no reason to keep `paperless-ng`, which won't be developed any further.
- Rename `services.paperless-ng` to `services.paperless` and use package `paperless-ngx`.
  This is fully backwards compatible and works seamlessly for existing users of `paperless-ng`.
  **Edit:** To [simplify the service](https://github.com/NixOS/nixpkgs/pull/168288#discussion_r848465902), I've introduced a [minor breaking change](https://github.com/NixOS/nixpkgs/blob/0977227b8622daccb5a8435d3d4216eedf845973/nixos/doc/manual/release-notes/rl-2205.section.md).

#### Things done

- [x] built `nixosTests.paperless` on `x86_64-linux`
- [x] manually tested the migration from an instance of `services.paperless-ng` to `services.paperless` (`paperless-ngx`)

lukegb committed:
paperless-ngx: init at 1.6.0

erikarvstedt committed:
paperless-ng: delete pkg, add alias to `paperless-ngx`

erikarvstedt committed:
nixos/paperless-ng: minor improvments

Service:
- Fix misleading comment:
  We could in fact implement password copying as a preStart script by
  amending BindReadOnlyPaths, but adding an extra service is simpler.

Test:
- Add more detailed subtest names
- Simplify date check

erikarvstedt committed:
nixos/paperless-ng: fix /proc access for service

This allows django-q to set the number of workers according to the CPU
core count and to show memory stats via `manage.py qmonitor`.
This also fixes a non-critical psutil error in paperless-ng-server.

ambroisie commented on nixos/tests/paperless.nix:
```diff
@@ -1,32 +1,30 @@
-    machine.wait_for_unit("paperless-ng-consumer.service")
-
-    with subtest("Add a document via the file system"):
+    with subtest("Create test doc"):
```
I think this is a regression from the previous commit?

erikarvstedt replied:
Thanks for catching this. History rewrites across renames can be a bit tricky.

ambroisie commented on nixos/tests/paperless.nix:
```diff
@@ -1,32 +1,30 @@
         machine.wait_until_succeeds("curl -fs localhost:28981")
 
-    with subtest("Add a document via the web interface"):
+    with subtest("Create web test doc"):
```
Ditto.

ambroisie commented on nixos/tests/paperless.nix:
```diff
@@ -37,8 +35,11 @@ import ./make-test-python.nix ({ lib, ... }: {
+        )
+        assert "2005-10-16" in machine.succeed(
+            "curl -u admin:admin -fs localhost:28981/api/documents/ | jq '.results | .[1] | .created'"
+        )
```
Ditto.

ambroisie reviewed (commented):
Other than a tiny issue in the test, LGTM.

Flakebi commented on pkgs/applications/office/paperless-ngx/default.nix:
```diff
@@ -43,30 +33,16 @@ let
-
   format = "other";
 
-  # Make bind address configurable
-  postPatch = ''
-    substituteInPlace gunicorn.conf.py --replace "bind = '0.0.0.0:8000'" ""
-  '';
-
```
Was this removed on purpose? I think the address is fixed to `0.0.0.0` by default

erikarvstedt replied:
Sorry for not explaining this.
Yes, I've removed it because it's not needed (and maybe never was).
You can see that in the NixOS test `paperless.port` is correctly bound, although env var `PAPERLESS_PORT` is never set by the service. This proves that `GUNICORN_CMD_ARGS` overrides the `bind` var in `gunicorn.conf.py`.

You can also check that only localhost is bound, like so:
```bash
config='{
  containers.tmp = {
    extra.addressPrefix = "10.30.0";
    config.services.paperless-ng.enable = true;
  };
}'
# Requires extra-container 0.9
extra-container shell --nixpkgs-path /path/to/nixpkgs -E "$config" --run bash -c 'sleep 1; c netstat -nltp'
# Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
# tcp        0      0 127.0.0.1:28981         0.0.0.0:*               LISTEN      422/python3.9       
#                     ^^^^^^^^^
```

gador commented:
Result of `nixpkgs-review pr 168288` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>paperless-ngx</li>
  </ul>
</details>

06kellyjac commented:
This should come with a notice in the release docs :)

mweinelt reviewed (commented):
Looks pretty good overall. Two remarks regarding backwards compat.

erikarvstedt committed:
nixos/paperless-ng: simplify redis logic

- Rename `hasCustomRedis` -> `enableRedis`
- Extract var `redisServer`

erikarvstedt committed:
nixos/paperless-ng: rename to `paperless`, use `paperless-ngx` package

Use this opportunity to rename `paperless-(ng)-server` to `paperless-scheduler`.
This follows the naming scheme in the paperless repo and docs.

erikarvstedt commented on nixos/modules/services/misc/paperless.nix:
```diff
@@ -2,11 +2,15 @@
+
   hasCustomRedis = hasAttr "PAPERLESS_REDIS" cfg.extraConfig;
+  redisServerName = if hasOldSystemState then "paperless-ng" else "paperless";
+  redisServer = config.services.redis.servers.${redisServerName};
```
Ah, it turned out that it doesn't. Thanks!

erikarvstedt commented on nixos/modules/services/misc/paperless.nix:
```diff
@@ -221,15 +224,18 @@ in
+        ln -sf ${manage} ${cfg.dataDir}/paperless-manage
+        ${optionalString hasOldSystemState ''
+          ln -sf ${manage} ${cfg.dataDir}/paperless-ng-manage
+        ''}
```
Scripts using the old path will fail.
But because `hasOldSystemState` is no longer needed for redis, we should make a minor backwards-incompatible change and remove the alias. I've noted this in the release notes.
